### PR TITLE
feat: support section name in policies, and add tests for all attachment types.

### PR DESCRIPTION
**Pull request by yuval-k**
# Description

Handle attachment edge cases, and allow policies to target section names.

# Change Type

```
/kind feature
```

# Changelog

```release-note
Allow policies to targetRef using section names.
```

# Additional Notes

Behavior after this PR:
- Route policies never attach to the envoy-listener (i.e. filters on the listener are disabled; so that without a per-filter route config they do nothing
- Route Policies that target ref a gateway will go on the route configuration
- Route policies that target a section will go on:
  - for http - on the virtual host
  - for https - on the route configuration 

Main implication is that HCM plugins can't attach to http plain text listeners by section name, because the http filter chain is shared. (But does work on tls).
Option to address this is to remove section name from the http listener policy.

## Additional changes

- Updated traffic policy to place filters on the specific filter chain.
- Made `getTargetingPolicies` to not return policies without a section name when a section name is requested. instead we use envoy's hierarchy to get the same data plane behaviour.

### CODE REVIEW

**Comment by yuval-k** [comment]
i noticed some multi listener tests are flakey due to ordering; i'll fix that; but otherwise ready for review

File: internal/kgateway/translator/irtranslator/route.go [hunk 1]
```diff
@@ -37,17 +38,35 @@ type httpRouteConfigurationTranslator struct {
 }
 
 func (h *httpRouteConfigurationTranslator) ComputeRouteConfiguration(ctx context.Context, vhosts []*ir.VirtualHost) *envoy_config_route_v3.RouteConfiguration {
+	attachedPoliciesSlice := []ir.AttachedPolicies{
+		h.gw.AttachedHttpPolicies,
+		h.attachedPolicies,
+	}
```
Comment (shashankram) [inline internal/kgateway/translator/irtranslator/route.go:44]: Could you pre-merge the policies per GK here so that it is easier to invoke a Merge() on them within the loop
```
	attachedPolicies := ir.AttachedPolicies{
		Policies: map[schema.GroupKind][]ir.PolicyAtt{},
	}
       h.gw.AttachedHttpPolicies.Append(h.gw.AttachedHttpPolicies, h.attachedPolicies)
```

File: internal/kgateway/translator/irtranslator/route.go [hunk 2]
```diff
@@ -37,17 +38,35 @@ type httpRouteConfigurationTranslator struct {
...
+	typedPerFilterConfigRoute := ir.TypedFilterConfigMap(map[string]proto.Message{})
+
+	for _, attachedPolicies := range attachedPoliciesSlice {
+		for gk, pols := range attachedPolicies.Policies {
+			pass := h.PluginPass[gk]
+			if pass == nil {
+				// TODO: user error - they attached a non http policy
+				continue
+			}
+			for _, pol := range pols {
```
Comment (shashankram) [inline internal/kgateway/translator/irtranslator/route.go:57]: Do you need to MergePolicies when `pass.MergePolicies != nil`?

File: internal/kgateway/translator/irtranslator/route.go [hunk 3]
```diff
@@ -37,17 +38,34 @@ type httpRouteConfigurationTranslator struct {
 }
 
 func (h *httpRouteConfigurationTranslator) ComputeRouteConfiguration(ctx context.Context, vhosts []*ir.VirtualHost) *envoy_config_route_v3.RouteConfiguration {
+
+	attachedPolicies := ir.AttachedPolicies{
+		Policies: map[schema.GroupKind][]ir.PolicyAtt{},
+	}
+	attachedPolicies.Append(h.gw.AttachedHttpPolicies, h.attachedPolicies)
```
Comment (shashankram) [inline internal/kgateway/translator/irtranslator/route.go:44]: nit: a comment on relative priority between the 2 would clarify the reason for the ordering here

File: internal/kgateway/translator/gateway/testutils/outputs/delegation/traffic_policy.yaml [hunk 4]
```diff
@@ -10,7 +10,8 @@ Listeners:
       typedConfig:
         '@type': type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
         httpFilters:
-        - name: transformation
+        - disabled: true
```
Comment (shashankram) [inline internal/kgateway/translator/gateway/testutils/outputs/delegation/traffic_policy.yaml:13]: Is `disabled: true` expected here and in other outputs?

File: internal/kgateway/setup/testdata/standard/extauth-gw-and-route-disable-out.yaml [hunk 5]
```diff
@@ -58,7 +58,8 @@ listeners:
             - metadataNamespace: dev.kgateway.disable_ext_auth
               value:
                 extauth_disable: true
-        - name: ext_auth/gwtest/basic-extauth
+        - disabled: true
```
[the remark this answers is not in the document]
Comment (yuval-k) [reply internal/kgateway/setup/testdata/standard/extauth-gw-and-route-disable-out.yaml:61]: i modified the traffic policy to always disable the filter on the listener. see additional notes on PR description.
I'm trying to get a consistent way to apply filters, so we don't need this `fromlistener` distinction

File: internal/kgateway/translator/gateway/testutils/outputs/delegation/traffic_policy.yaml (hunk 4 above)
Comment (shashankram) [reply internal/kgateway/translator/gateway/testutils/outputs/delegation/traffic_policy.yaml:13]: Ah, the file diff was condensed by GH so missed it

**Review by shashankram — APPROVED** [review]

**Review by shashankram — APPROVED** [review]
